File: 9_smoke_basin/smoke_basin.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strings"
)

func getInputs(filename string) ([][]int, error) {
	data, err := os.ReadFile(filename)
	if err != nil {
		return nil, err
	}
	lines := strings.Split(string(data), "\n")
	mapData := make([][]int, len(lines))
	for i := range mapData {
		mapData[i] = make([]int, len(lines[i]))
	}
	for i := range lines {
		for j := range lines[i] {
			mapData[i][j] = int(lines[i][j] - '0')
		}
	}
	return mapData, nil
}

func sumLowPointRiskLevel(mapData [][]int) int {
	riskLevel := 0
	for i := range mapData {
		for j := range mapData[i] {
			if i-1 >= 0 && mapData[i-1][j] <= mapData[i][j] {
				continue
			}
			if i+1 < len(mapData) && mapData[i+1][j] <= mapData[i][j] {
				continue
			}
			if j-1 >= 0 && mapData[i][j-1] <= mapData[i][j] {
				continue
			}
			if j+1 < len(mapData[i]) && mapData[i][j+1] <= mapData[i][j] {
				continue
			}
			riskLevel += (mapData[i][j] + 1)
		}
	}
	return riskLevel
}

func initMapData(mapData [][]int) {
	for i := range mapData {
		for j := range mapData[i] {
			if mapData[i][j] == 9 {
				mapData[i][j] = 1
			} else {
				mapData[i][j] = 0
			}
		}
	}
}

func tagBasin(mapData [][]int, i, j int) int {
	if mapData[i][j] != 0 {
		return 0
	}
	mapData[i][j] = 1
	res := 1
	if i-1 >= 0 {
		res += tagBasin(mapData, i-1, j)
	}
	if i+1 < len(mapData) {
		res += tagBasin(mapData, i+1, j)
	}
	if j-1 >= 0 {
		res += tagBasin(mapData, i, j-1)
	}
	if j+1 < len(mapData[i]) {
		res += tagBasin(mapData, i, j+1)
	}
	return res
}

func multiplicationThreeLargestBasins(mapData [][]int) int {
	initMapData(mapData)
	var topThree []int
	for i := range mapData {
		for j := range mapData[i] {
			if mapData[i][j] != 0 {
				continue
			}
			sizeBasin := tagBasin(mapData, i, j)
			if len(topThree) < 3 {
				topThree = append(topThree, sizeBasin)
			} else {
				sort.Ints(topThree)
				if sizeBasin > topThree[0] {
					topThree[0] = sizeBasin
				}
			}
		}
	}
	return topThree[0] * topThree[1] * topThree[2]
}

func main() {
	mapData, err := getInputs("input.txt")
	if err != nil {
		fmt.Println(err)
		return
	}
	fmt.Printf("Sum of risk level: %d\n", sumLowPointRiskLevel(mapData))
	fmt.Printf("Multiplication of the sizes of top 3 basins %d\n",
		multiplicationThreeLargestBasins(mapData))
}

File: 9_smoke_basin/smoke_basin.py
import math
from typing import List


def get_inputs(file_name: str) -> List[List[int]]:
    with open(file_name) as f:
        data = f.readlines()
    return [[int(c) for c in line.strip()] for line in data]


def sum_risk_level_of_low_points(map_data: List[List[int]]) -> int:
    sum_risk_level = 0
    for i in range(len(map_data)):
        for j in range(len(map_data[i])):
            if i - 1 >= 0 and map_data[i - 1][j] <= map_data[i][j]:
                continue
            if i + 1 < len(map_data) and map_data[i + 1][j] <= map_data[i][j]:
                continue
            if j - 1 >= 0 and map_data[i][j - 1] <= map_data[i][j]:
                continue
            if j + 1 < len(map_data[i]) and map_data[i][j + 1] <= map_data[i][j]:
                continue
            sum_risk_level += (map_data[i][j] + 1)
    return sum_risk_level


def init_map_data(map_data: List[List[int]]) -> None:
    for i in range(len(map_data)):
        for j in range(len(map_data[i])):
            if map_data[i][j] == 9:
                map_data[i][j] = 1
            else:
                map_data[i][j] = 0


def tag_basin(map_data: List[List[int]], i: int, j: int) -> int:
    if map_data[i][j] != 0:
        return 0
    map_data[i][j] = 1
    res = 1
    if i - 1 >= 0:
        res += tag_basin(map_data, i - 1, j)
    if i + 1 < len(map_data):
        res += tag_basin(map_data, i + 1, j)
    if j - 1 >= 0:
        res += tag_basin(map_data, i, j - 1)
    if j + 1 < len(map_data[i]):
        res += tag_basin(map_data, i, j + 1)
    return res


def multiplication_of_three_largest_basins(map_data: List[List[int]]) -> int:
    init_map_data(map_data)
    top_three: List[int] = []
    for i in range(len(map_data)):
        for j in range(len(map_data[i])):
            if map_data[i][j] != 0:
                continue
            size_basin = tag_basin(map_data, i, j)
            if len(top_three) < 3:
                top_three.append(size_basin)
            else:
                top_three = sorted(top_three)
                if size_basin > top_three[0]:
                    top_three[0] = size_basin
    return math.prod(top_three)


if __name__ == '__main__':
    map_data = get_inputs('input.txt')
    print(f'Sum of risk level: {sum_risk_level_of_low_points(map_data)}')
    print('Multiplication of the sizes of top 3 basins: {}'
          .format(multiplication_of_three_largest_basins(map_data)))
